lib/src/model_utils.dart

```diff
@@ -166,3 +169,150 @@ String crossdartifySource(
   }
   return newSource;
 }
+
+/// An UnmodifiableListView that computes equality and hashCode based on the
+/// equality and hashCode of its contained objects.
+class HashableList extends UnmodifiableListView<dynamic> {
+  HashableList(Iterable<dynamic> iterable) : super(iterable);
+
+  @override
+  bool operator ==(other) {
+    if (this.length == other.length) {
+      for (var index = 0; index < length; ++index) {
+        if (this[index] != other[index]) return false;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  @override
+  get hashCode => hashObjects(this);
+}
+
+/// A type alias for [Tuple2]`<Function, HashableList>`.
+class _MemoKey extends Tuple2<Function, HashableList> {
+  _MemoKey(Function f, HashableList l) : super(f, l) {}
+}
+
+/// Extend or use as a mixin to track object-specific cached values, or
+/// instantiate directly to track other values.
+class Memoizer {
+  /// Map of a function and its positional parameters (if any), to a value.
+  Map<_MemoKey, dynamic> _memoizationTable;
+
+  Memoizer() {
+    invalidateMemos();
+  }
+
+  /// Reset the memoization table, forcing calls of the underlying functions.
+  void invalidateMemos() {
+    _memoizationTable = new Map();
+  }
+
+  // Never write this if statement again.
+  R _getOrUpdateCache<R>(R Function() f, _MemoKey key) {
+    if (!_memoizationTable.containsKey(key)) {
```

Comment:
This should be more performant:
```dart
R _getOrUpdateCache<R>(R Function() f, _MemoKey key) {
  return _memoizationTable.putIfAbsent(key, f);
}
```
...because `putIfAbsent` only needs to hash the key once.

(Note that in ordinary circumstances this benefit is offset by the expense of creating the closure to pass to `putIfAbsent`, but in this particular case we're paying that expense anyhow).

Reply:
Actually if you look at the definition of putIfAbsent you'll see that it already has the behavior we want:
```dart
  /**                                                                                                                                                                                                                                                                                     
   * Look up the value of [key], or add a new value if it isn't there.                                                                                                                                                                                                                    
   *                                                                                                                                                                                                                                                                                      
   * Returns the value associated to [key], if there is one.                                                                                                                                                                                                                              
   * Otherwise calls [ifAbsent] to get a new value, associates [key] to                                                                                                                                                                                                                   
   * that value, and then returns the new value.                                                                                                                                                                                                                                          
   *                                                                                                                                                                                                                                                                                      
   *     Map<String, int> scores = {'Bob': 36};                                                                                                                                                                                                                                           
   *     for (var key in ['Bob', 'Rohan', 'Sophena']) {                                                                                                                                                                                                                                   
   *       scores.putIfAbsent(key, () => key.length);                                                                                                                                                                                                                                     
   *     }                                                                                                                                                                                                                                                                                
   *     scores['Bob'];      // 36                                                                                                                                                                                                                                                        
   *     scores['Rohan'];    //  5                                                                                                                                                                                                                                                        
   *     scores['Sophena'];  //  7                                                                                                                                                                                                                                                        
   *                                                                                                                                                                                                                                                                                      
   * Calling [ifAbsent] must not add or remove keys from the map.                                                                                                                                                                                                                         
   */
  V putIfAbsent(K key, V ifAbsent());
```
So you should be able to do this directly:
```dart
R _getOrUpdateCache<R>(R Function() f, _MemoKey key) {
  return _memoizationTable.putIfAbsent(key, f); // Note lack of `()` after `f`
}
```

Reply:
**facepalm**  I don't know why I thought putAbsent worked the other way.  Maybe because of Python's dict.setdefault.  Fixed.